File: 2021년/8월/0809 종이의 개수.cpp
#include <iostream>
using namespace std;

int maps[2187][2187];
int mi = 0, ze = 0, on = 0;

void paper(int m, int x, int y)
{
	int number = maps[x][y];
	for(int i=x;i<m+x;i++)
	{
		for(int j=y;j<m+y;j++)
		{
			if(number != maps[i][j])
			{
				for(int k=0;k<3;k++)
				{
					paper(m/3, x+m/3*k, y);
					paper(m/3, x+m/3*k, y+m/3);
					paper(m/3, x+m/3*k, y+m/3*2);
				}
				return;
			}
		}
	}
	if(number == -1) mi++;
	else if(number == 0) ze++;
	else on++;
}

int main()
{
	int n;
	cin >> n;
	for(int i=0;i<n;i++)
	{
		for(int j=0;j<n;j++) cin >> maps[i][j];
	}
	
	paper(n, 0, 0);
	
	cout << mi << endl << ze << endl << on;
}

File: 2021년/8월/0809 종이의 개수.py
# 백준_1780_종이의 개수_분할 정복_실버 2

import sys
input = sys.stdin.readline


def paper(m, x, y):
    global minus, zero, one
    number = maps[x][y]
    for i in range(x, x+m):
        for j in range(y, y+m):
            if number != maps[i][j]:
                for k in range(3):
                    paper(m//3, x+k*m//3, y)
                    paper(m//3, x+k*m//3, y+m//3)
                    paper(m//3, x+k*m//3, y+m//3*2)
                return

    if number == -1:
        minus += 1
    elif number == 0:
        zero += 1
    else:
        one += 1


n = int(input())
maps = []
for _ in range(n):
    maps.append(list(map(int, input().split())))

minus, zero, one = 0, 0, 0

paper(n, 0, 0)

print(minus)
print(zero)
print(one)
